app/src/main/java/org/mozilla/reference/browser/browser/MainContainerFragment.kt

```diff
@@ -0,0 +1,310 @@
+    private val readerViewAppearanceButton: FloatingActionButton
+        get() = requireView().findViewById(R.id.readerViewAppearanceButton)
+
+    // Features
+    private val webExtToolbarFeature = ViewBoundFeatureWrapper<WebExtensionToolbarFeature>()
+    private val toolbarIntegration = ViewBoundFeatureWrapper<ToolbarIntegration>()
+    private val readerViewFeature = ViewBoundFeatureWrapper<ReaderViewIntegration>()
+
+    private val backButtonHandlers: List<ViewBoundFeatureWrapper<*>> = listOf(
+        readerViewFeature,
+        toolbarIntegration,
+    )
+
+    private val translationYOnDrawFetcher by lazy {
+        TranslationYOnDrawFetcher(
+            toolbar,
+            onDraw = { offset -> viewModel.updateToolbarOffset(offset) }
+        )
+    }
+
+    private val sessionId: String? by lazy { arguments?.getString(SESSION_ID) }
+
+    private val navHost by lazy {
+        childFragmentManager.findFragmentById(R.id.container) as NavHostFragment
+    }
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?,
+    ): View = inflater.inflate(R.layout.fragment_main_container, container, false)
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        setupToolbar(view)
+        setupAwesomeBar()
+        setupTabsToolbarFeature()
+        setupWebExtToolbarFeature(view)
+        setupReaderMode(view)
+
+        setupFragmentResultListener()
+        observeNavigationEvents()
+    }
+
+    override fun onBackPressed(): Boolean {
+        logger.debug("onBackPressed")
+        // Checks if any of the features handled the back press
+        // and if any child fragments that are UserInteractionHandlers handled it
+        return backButtonHandlers.any { it.onBackPressed() } ||
+                navHost.childFragmentManager.fragments
+                    .filterIsInstance<UserInteractionHandler>()
+                    .any {
+                        it.onBackPressed()
+                    }
+    }
+
+    override fun onDestroyView() {
+        toolbar.viewTreeObserver.removeOnDrawListener(translationYOnDrawFetcher)
+        super.onDestroyView()
+    }
+
+    private fun setupToolbar(view: View) {
+        toolbar.viewTreeObserver.addOnDrawListener(translationYOnDrawFetcher)
+
+        //        toolbar.viewTreeObserver.addOnPreDrawListener {
+        //            val currentTranslationY = toolbar.translationY
+        //            viewModel.updateToolbarOffset(currentTranslationY)
+        //            logger.debug("onPreDraw: $currentTranslationY")
+        //            true
+        //        }
+
+        (toolbar.layoutParams as? CoordinatorLayout.LayoutParams)?.apply {
+            behavior = EngineViewScrollingBehavior(
+                view.context,
+                null,
+                ViewPosition.BOTTOM,
+            )
+        }
```

Comment:
As a small note, we're already working with Gecko - APZ to expose the scroll data through GeckoView and hoping in a few months we'll be able to do this more elegantly - through the usual nested scrolling framework.
But for now it's great that translating the toolbar based on a child fragment is supported by `EngineViewScrollingBehavior`.